bin/racket_exercise_generator.py

```diff
@@ -0,0 +1,553 @@
+    config_data["files"]["test"] = [f"{exercise_name}-test.rkt"]
+    config_data["files"]["solution"] = [f"{exercise_name}.rkt"]
+    config_data["files"]["example"] = [".meta/example.rkt"]
+    config_data["authors"] = [author]
+    config_data["contributors"] = []
+
+    with open(f"{TARGET}/{exercise_name}/.meta/config.json", "w") as file:
+        # Encode into a string in json format and write to file
+        # TODO: Use configlet instead? What about the author?
+        file.write(json.dumps(config_data, cls=CustomJSONEncoder, indent=3))
+        file.write("\n")
+
+
+def create_instructions(exercise_name, prob_spec_exercise):
+    """
+    Auto-generates the .docs/instructions.md file
+
+    Parameter exercise_name: Name of the exercise to be generated.
+    Precondition: exercise_name is a string of a valid exercise.
+
+    Parameter prob_spec_exercise: A filepath to the location of the exercise folder
+    in the problem-specifications repository.
+    Precondition: prob_spec_exercise is a string of a valid filepath.
+    """
+    input_file = f"{prob_spec_exercise}/description.md"
+    output_file = f"{TARGET}/{exercise_name}/.docs/instructions.md"
+    with open(input_file, "r", encoding="utf-8") as read_from:
+        with open(output_file, "w", encoding="utf-8") as write_to:
+            # Replace first line with "# Instructions\n" during copy process
+            write_to.write(
+                "# Instructions\n" + "\n".join(read_from.read().split("\n")[1:])
+            )
+
+
+def create_test_example_solution_files(exercise_name, prob_spec_exercise):
+    """
+    Auto-generates the test file.
+
+    Function creates the test file in its own right, but also calls the
+    create_example_and_solution_files function.  This function also creates
+    the parameters to feed into the create_example_and_solution_files function.
+
+    Parameter exercise_name: Name of the exercise to be generated.
+    Precondition: exercise_name is a string of a valid exercise.
+
+    Parameter prob_spec_exercise: A filepath to the location of the exercise folder
+    in the problem-specifications repository.
+    Precondition: prob_spec_exercise is a string of a valid filepath.
+    """
+    data = None
+    with open(f"{prob_spec_exercise}/canonical-data.json") as file:
+        data = json.load(file)
+
+    # Boilerplate test code.  Multiline docstring format used to maintain
+    # correct indentation and to increase readability.
+    exercise_string = """#lang racket/base
+
+(require "{0}.rkt")
+
+(module+ test
+  (require rackunit rackunit/text-ui)
+""".format(
+        exercise_name
+    )
+
+    # Check if some tests expect errors
+    expect_errors = any(
+        [
+            "error" in case["expected"].keys()
+            for case in data["cases"]
+            if isinstance(case["expected"], dict)
+        ]
+    )
+
+    # If so, add the definition of the helper function to test returned errors
+    if expect_errors:
```

Comment:
I think we want to avoid checking exception messages because some students may use contracts instead so that won’t work.

Reply:
Ok, I based myself on `collatz-conjecture-test.rkt`.

I guess using `exn:fail?` is preferred then?